### PR TITLE
feat(core): add parallel LWE PFPKS and LWE KS

### PR DESCRIPTION
**Pull request by IceTDrinker**
### PR content/description

Parallel version of the keyswitching algorithms have been added

### Check-list:

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [x] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: tfhe/src/core_crypto/entities/lwe_private_functional_packing_keyswitch_key.rs [hunk 1]
```diff
@@ -378,7 +387,7 @@ impl<Scalar: UnsignedInteger, C: ContainerMut<Element = Scalar>> ContiguousEntit
 
     // At the moment it does not make sense to return "sub" packing keyswitch keys. So we use a
     // dummy placeholder type here.
```
Comment (mayeul-zama) [inline tfhe/src/core_crypto/entities/lwe_private_functional_packing_keyswitch_key.rs:389]: Update/remove comment

File: tfhe/src/core_crypto/entities/lwe_private_functional_packing_keyswitch_key.rs [hunk 2]
```diff
@@ -362,9 +368,12 @@ impl<Scalar: UnsignedInteger, C: Container<Element = Scalar>> ContiguousEntityCo
     /// Unimplemented for [`LwePrivateFunctionalPackingKeyswitchKey`]. At the moment it does not
     /// make sense to return "sub" packing keyswitch keys.
```
Comment (mayeul-zama) [inline tfhe/src/core_crypto/entities/lwe_private_functional_packing_keyswitch_key.rs:369]: Update/remove comment

File: tfhe/src/core_crypto/entities/lwe_private_functional_packing_keyswitch_key.rs [hunk 3]
```diff
@@ -339,11 +345,11 @@ impl<Scalar: UnsignedInteger, C: Container<Element = Scalar>> ContiguousEntityCo
     where
         Self: 'this;
 
-    type SelfViewMetadata = ();
+    type SelfViewMetadata = LwePrivateFunctionalPackingKeyswitchKeyCreationMetadata<Self::Element>;
 
     // At the moment it does not make sense to return "sub" packing keyswitch keys. So we use a
     // dummy placeholder type here.
```
Comment (mayeul-zama) [inline tfhe/src/core_crypto/entities/lwe_private_functional_packing_keyswitch_key.rs:351]: Update/remove comment

File: tfhe/src/core_crypto/entities/lwe_keyswitch_key.rs [hunk 4]
```diff
@@ -373,9 +402,11 @@ impl<Scalar: UnsignedInteger, C: Container<Element = Scalar>> ContiguousEntityCo
     /// Unimplemented for [`LweKeyswitchKey`]. At the moment it does not make sense to
     /// return "sub" keyswitch keys.
```
Comment (mayeul-zama) [inline tfhe/src/core_crypto/entities/lwe_keyswitch_key.rs:403]: Update/remove comment

File: tfhe/src/core_crypto/entities/lwe_keyswitch_key.rs [hunk 5]
```diff
@@ -354,11 +383,11 @@ impl<Scalar: UnsignedInteger, C: Container<Element = Scalar>> ContiguousEntityCo
     where
         Self: 'this;
 
-    type SelfViewMetadata = ();
+    type SelfViewMetadata = LweKeyswitchKeyCreationMetadata<Self::Element>;
 
     // At the moment it does not make sense to return "sub" keyswitch keys. So we use a dummy
     // placeholder type here.
```
Comment (mayeul-zama) [inline tfhe/src/core_crypto/entities/lwe_keyswitch_key.rs:389]: Update/remove comment

**Review by mayeul-zama — APPROVED** [review]

File: tfhe/src/core_crypto/fft_impl/fft64/crypto/wop_pbs/mod.rs [hunk 6]
```diff
@@ -309,25 +310,23 @@ pub fn circuit_bootstrap_boolean<Scalar: UnsignedTorus + CastInto<usize>>(
...
         LweCiphertext::from_container(&mut *lwe_out_bs_buffer_data, lwe_in.ciphertext_modulus());
 
-    // Output for every pfksk that come from the output GGSW
-    let mut glwe_out_pfksk_buffer = ggsw_out.as_mut_glwe_list();
-
-    let mut out_pfksk_buffer_iter = glwe_out_pfksk_buffer.iter_mut();
-
-    for decomposition_level in (1..=level_cbs.0).map(DecompositionLevelCount) {
+    for (mut decomposition_level, mut ggsw_level_matrix) in ggsw_out.iter_mut().enumerate() {
+        decomposition_level += 1;
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/core_crypto/fft_impl/fft64/crypto/wop_pbs/mod.rs:314]: fair enough and yes the indexing starting from 1 is to match the decomposition logic as it was written

File: tfhe/src/core_crypto/entities/lwe_private_functional_packing_keyswitch_key.rs (hunk 1 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/entities/lwe_private_functional_packing_keyswitch_key.rs:389]: arf, thought I removed that 🤔

**Comment by IceTDrinker** [comment]
relaunching to have the CI on the latest code change
@slab-ci cpu_test
@slab-ci cpu_integer_test
@slab-ci cpu_multi_bit_test
@slab-ci cpu_wasm_test

**Pull request merged**